pkg/kp/dsstore/consul_store.go

```diff
@@ -0,0 +1,207 @@
+
+	// TODO: measure whether retries are is important in practice
+	for i := 0; i < s.retries; i++ {
+		if _, ok := err.(CASError); ok {
+			ds, err = s.innerCreate(manifest, minHealth, name, nodeSelector, podID)
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		return fields.DaemonSet{}, err
+	}
+	return ds, nil
+}
+
+// these parts of Create may require a retry
+func (s *consulStore) innerCreate(
+	manifest pods.Manifest,
+	minHealth int,
+	name fields.ClusterName,
+	nodeSelector klabels.Selector,
+	podID types.PodID,
+) (fields.DaemonSet, error) {
+	id := fields.ID(uuid.New())
+	dsPath, err := s.dsPath(id)
+	if err != nil {
+		return fields.DaemonSet{}, err
+	}
+	ds := fields.DaemonSet{
+		ID:           id,
+		Disabled:     false,
+		Manifest:     manifest,
+		MinHealth:    minHealth,
+		Name:         name,
+		NodeSelector: nodeSelector,
+		PodID:        podID,
+	}
+	// Marshals ds into []bytes using overloaded MarshalJSON
+	rawDS, err := json.Marshal(ds)
+	if err != nil {
+		return fields.DaemonSet{}, util.Errorf("Could not marshal DS as json: %s", err)
+	}
+
+	success, _, err := s.kv.CAS(&api.KVPair{
+		Key:         dsPath,
+		Value:       rawDS,
+		ModifyIndex: 0,
+	}, nil)
+
+	if err != nil {
+		return fields.DaemonSet{}, consulutil.NewKVError("cas", dsPath, err)
+	}
+	if !success {
+		return fields.DaemonSet{}, CASError(dsPath)
+	}
+	return ds, nil
+}
+
+func (s *consulStore) Get(id fields.ID) (fields.DaemonSet, *api.QueryMeta, error) {
+	dsPath, err := s.dsPath(id)
+	var metadata *api.QueryMeta
+
+	if err != nil {
+		return fields.DaemonSet{}, metadata, err
+	}
+
+	kvp, metadata, err := s.kv.Get(dsPath, nil)
+	if err != nil {
+		return fields.DaemonSet{}, metadata, consulutil.NewKVError("get", dsPath, err)
+	}
+	if metadata == nil {
+		// no metadata returned
+		return fields.DaemonSet{}, metadata, errors.New("No metadata found")
+	}
+	if kvp == nil {
+		// ID didn't exist
+		return fields.DaemonSet{}, metadata, NoDaemonSet
+	}
+
+	ds, err := kvpToDS(kvp)
+	if err != nil {
+		return fields.DaemonSet{}, metadata, err
+	}
+	return ds, metadata, err
+}
+
+func (s *consulStore) List() ([]fields.DaemonSet, error) {
+	listed, _, err := s.kv.List(dsTree+"/", nil)
+	if err != nil {
+		return nil, consulutil.NewKVError("list", dsTree+"/", err)
+	}
+	return kvpsToDSs(listed)
+}
+
+func checkManifestPodID(dsPodID types.PodID, manifest pods.Manifest) error {
+	if dsPodID == "" {
+		return util.Errorf("Daemon set must have a pod id")
+	}
+	if manifest.ID() == "" {
+		return util.Errorf("Daemon set manifest must have a pod id")
+	}
+	if dsPodID != manifest.ID() {
+		return util.Errorf("Daemon set pod id must match manifest pod id. Wanted '%s', got '%s'", dsPodID, manifest.ID())
+	}
+	return nil
+}
+
+func (s *consulStore) dsPath(dsID fields.ID) (string, error) {
+	if dsID == "" {
+		return "", util.Errorf("Path requested for empty DS id")
+	}
+
+	return path.Join(dsTree, string(dsID)), nil
```

Comment:
`dsID.String()`. That way if that type ever changes we won't be blindly casting it to string, the compiler will warn us